### ISSUE TITLE
Unload lora specific lora weight

### ISSUE THREAD

**Issue by AMohamedAakhil** [huggingface/diffusers#9325]
I've had trouble loading a few loras, and then unloading only a select few to free up more memory, but we only support unload_lora_weights() currently, which unloads all the loras - so I have to reload the loras I want again which is time consuming. 

I would like a solution where we could specify the adapter_name associated with the lora_weight and unload that specific lora weight alone.

**Comment by asomoza** [comment]
let's keep this one open to track this feature request. I've seen a couple more issues with the same.

**Comment by rickstaa** [comment]
@AMohamedAakhil, @asomoza, we successfully implemented dynamic loading for our project by leveraging a combination of `load_lora_weights`, `delete_adapters`, `set_adapters`, `disable_lora`, `torch.cuda.empty_cache()`, and a caching map. You can check out the code [here](https://github.com/livepeer/ai-worker/blob/0d96ac0f6f69835888bad271093865928bba7cff/runner/app/pipelines/utils/utils.py#L217-L366). Hope that helps, if you have any questions let me know.

**Comment by sayakpaul** [comment]
@rickstaa would you like to maybe help us with a PR? That would be extremely valuable for us. If not, I can completely understand, I will take this up.

**Comment by sayakpaul** [comment]
@rickstaa sorry for the ping but just wanted to check.

**Comment by sayakpaul** [comment]
I have some updates here. 

The changes are relatively simple but there are things to be discussed detailed below. 

<details>
<summary>patch</summary>

```diff
diff --git a/src/diffusers/loaders/lora_base.py b/src/diffusers/loaders/lora_base.py
index 286d0a12b..e84343b26 100644
--- a/src/diffusers/loaders/lora_base.py
+++ b/src/diffusers/loaders/lora_base.py
@@ -361,7 +361,7 @@ class LoraBaseMixin:
         deprecate("_best_guess_weight_name", "0.35.0", deprecation_message)
         return _best_guess_weight_name(*args, **kwargs)
 
-    def unload_lora_weights(self):
+    def unload_lora_weights(self, adapter_names=None):
         """
         Unloads the LoRA parameters.
 
@@ -380,8 +380,9 @@ class LoraBaseMixin:
             model = getattr(self, component, None)
             if model is not None:
                 if issubclass(model.__class__, ModelMixin):
-                    model.unload_lora()
+                    model.unload_lora(adapter_names=adapter_names)
                 elif issubclass(model.__class__, PreTrainedModel):
+                    # todo: handle adapter_names
                     _remove_text_encoder_monkey_patch(model)
 
     def fuse_lora(
diff --git a/src/diffusers/loaders/peft.py b/src/diffusers/loaders/peft.py
index bf118c88b..8032f5eb9 100644
--- a/src/diffusers/loaders/peft.py
+++ b/src/diffusers/loaders/peft.py
@@ -599,15 +599,19 @@ class PeftAdapterMixin:
         if isinstance(module, BaseTunerLayer):
             module.unmerge()
 
-    def unload_lora(self):
+    def unload_lora(self, adapter_names=None):
         if not USE_PEFT_BACKEND:
             raise ValueError("PEFT backend is required for `unload_lora()`.")
 
         from ..utils import recurse_remove_peft_layers
 
-        recurse_remove_peft_layers(self)
-        if hasattr(self, "peft_config"):
-            del self.peft_config
+        if adapter_names is None:
+            recurse_remove_peft_layers(self)
+            if hasattr(self, "peft_config"):
+                del self.peft_config
+        else:
+            # TODO: add validation on `adapter_names` if they exist
+            self.delete_adapters(adapter_names=adapter_names)
 
     def disable_lora(self):
         """

```

</details>

And here is my simple testing script:

```py
from diffusers import DiffusionPipeline
import numpy as np
import torch 

pipe = DiffusionPipeline.from_pretrained("black-forest-labs/FLUX.1-dev", torch_dtype=torch.bfloat16).to("cuda")
pipe.load_lora_weights("XLabs-AI/flux-RealismLora", adapter_name="realism")
out_1 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images

pipe.load_lora_weights("ginipick/flux-lora-eric-cat", adapter_name="cat")
print(pipe.get_active_adapters())
out_2 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images
assert not np.allclose(out_1, out_2, atol=1e-4, rtol=1e-4)

pipe.set_adapters(["realism", "cat"])
print(pipe.get_active_adapters())
out_3 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images
assert not np.allclose(out_1, out_3, atol=1e-4, rtol=1e-4)
assert not np.allclose(out_2, out_3, atol=1e-4, rtol=1e-4)

pipe.set_adapters("cat")
print(pipe.get_active_adapters())
out_4 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images
assert not np.allclose(out_1, out_4, atol=1e-4, rtol=1e-4)
assert np.allclose(out_2, out_4, atol=1e-4, rtol=1e-4)

pipe.unload_lora_weights(adapter_names="realism")
print(pipe.get_active_adapters())
out_5 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images
assert np.allclose(out_4, out_5, atol=1e-4, rtol=1e-4)
assert np.allclose(out_2, out_5, atol=1e-4, rtol=1e-4)
```

@BenjaminBossan currently, this is just calling the [`delete_adapters()`](https://github.com/huggingface/diffusers/blob/6131a93b969f87d171148bd367fd9990d5a49b6b/src/diffusers/loaders/peft.py#L658) method as one can notice but that is probably not we want as it's not equivalent to unloading the weights associated with the provided `adapter_name` completely. 

Whereas [`recurse_remove_peft_layers()`](https://github.com/huggingface/diffusers/blob/6131a93b969f87d171148bd367fd9990d5a49b6b/src/diffusers/utils/peft_utils.py#L31) removes the layers recursively. What are your thoughts?

**Comment by sayakpaul** [comment]
I have another hackier approach that uses `setattr()`:

```py
from diffusers import DiffusionPipeline
from diffusers.utils import delete_adapter_layers
import numpy as np
import torch 

pipe = DiffusionPipeline.from_pretrained("black-forest-labs/FLUX.1-dev", torch_dtype=torch.bfloat16).to("cuda")
pipe.load_lora_weights("XLabs-AI/flux-RealismLora", adapter_name="realism")
out_1 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images

pipe.load_lora_weights("ginipick/flux-lora-eric-cat", adapter_name="cat")
assert pipe.get_active_adapters() == ["cat"]

out_2 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images
assert not np.allclose(out_1, out_2, atol=1e-4, rtol=1e-4)

pipe.set_adapters(["realism", "cat"])
assert pipe.get_active_adapters() == ["realism", "cat"]
out_3 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images
assert not np.allclose(out_1, out_3, atol=1e-4, rtol=1e-4)
assert not np.allclose(out_2, out_3, atol=1e-4, rtol=1e-4)

# unload specific adapter completely
state_dict = pipe.transformer.state_dict()
adapter_name_to_unload = "realism"
config = pipe.transformer.peft_config[adapter_name_to_unload]

for n, k in pipe.transformer.named_modules():
    if adapter_name_to_unload in n:
        setattr(pipe.transformer, n, None)

pipe.transformer.peft_config.pop(adapter_name_to_unload)
delete_adapter_layers(pipe.transformer, adapter_name_to_unload)
assert pipe.get_active_adapters() == ["cat"]
out_4 = pipe(prompt="a dog", num_inference_steps=5, generator=torch.manual_seed(0), output_type="np").images
assert not np.allclose(out_1, out_4, atol=1e-4, rtol=1e-4)
assert np.allclose(out_2, out_4, atol=1e-4, rtol=1e-4)
```

LMK.

**Comment by sayakpaul** [comment]
Sorry, forgot to tag @BenjaminBossan for the above two comments.

**Comment by BenjaminBossan** [comment]
> currently, this is just calling the [`delete_adapters()`](https://github.com/huggingface/diffusers/blob/6131a93b969f87d171148bd367fd9990d5a49b6b/src/diffusers/loaders/peft.py#L658) method as one can notice but that is probably not we want as it's not equivalent to unloading the weights associated with the provided `adapter_name` completely.
> 
> Whereas [`recurse_remove_peft_layers()`](https://github.com/huggingface/diffusers/blob/6131a93b969f87d171148bd367fd9990d5a49b6b/src/diffusers/utils/peft_utils.py#L31) removes the layers recursively. What are your thoughts?

Yes, I think we need to be careful to distinguish between 2 things. One thing is that we want to delete a specific adapter. After doing this with `delete_adapter`, the LoRA layer is still there. This is good if we have multiple adapters and only want to delete a specific one.

The other thing is if we want to completely get rid of the LoRA layers, meaning we just have normal torch layers left at the end. This is what `recurse_remove_peft_layers` does (and what `unload` does in PEFT). If more than one adapter is loaded, it will remove all of them. Therefore, for the initial request of allowing to delete specific adapters, there is no choice but to go the first route.

A design question for diffusers would be: If I delete the last adapter, do I want to automatically unload all LoRA layers? If yes, we could check after `delete_adapter` has been called on all layers whether there are any adapters left, and call `recurse_remove_peft_layers` in addition.

Btw when I checked these lines:

https://github.com/huggingface/diffusers/blob/6131a93b969f87d171148bd367fd9990d5a49b6b/src/diffusers/utils/peft_utils.py#L66-L90

we should be able to simplify it by calling `base_layer = module.get_base_layer()` instead of constructing new layers from scratch, [similar to what we do in PEFT](https://github.com/huggingface/peft/blob/ec92cdcc41fe1b141bfe1e0da69b38a7e601cc80/src/peft/tuners/lora/model.py#L515). This is working since PEFT v0.7.0.

**Comment by BenjaminBossan** [comment]
> Sounds good but WDYT about the approach in [#9325 (comment)](https://github.com/huggingface/diffusers/issues/9325#issuecomment-2528559682)?

> ```python
> for n, k in pipe.transformer.named_modules():
>     if adapter_name_to_unload in n:
>         setattr(pipe.transformer, n, None)
> 
> pipe.transformer.peft_config.pop(adapter_name_to_unload)
> delete_adapter_layers(pipe.transformer, adapter_name_to_unload)
> ```

Honestly, I'm not sure how/why this works :)

> I think this could be parameterized through a function argument and that way users exactly know what they are doing. WDYT?

That could be an option, but I'm not sure what the default should be. We could also forbid deleting the last remaining adapter and instead tell users to unload instead, just to throw another idea out there.

> But since we have the check for PEFT v0.6.0, it could add another version branch for 0.7.0, no? Or am I missing something?

You mean that we want to ensure that PEFT v0.6.0 still works? In that case, we can leave it as is and not complicate the code. I'm not sure if there is any plan to bump the minimal PEFT version. If yes, we could at least add a comment to change this once v0.6.0 is dropped.

**Comment by sayakpaul** [comment]
Okay, seems like `delete_adapters()` is the way to go here. Will open a PR!

> You mean that we want to ensure that PEFT v0.6.0 still works? In that case, we can leave it as is and not complicate the code. I'm not sure if there is any plan to bump the minimal PEFT version. If yes, we could at least add a comment to change this once v0.6.0 is dropped.

Sounds good, I will club this in my PR.

**Comment by sayakpaul** [comment]
Hmm, @BenjaminBossan I have been thinking about this. Given we're just repurposing `delete_adapters()` here to unload a specific LoRA, I wonder if it makes sense to add `adapter_names` to `unload_lora_weights()`. Users can always just use `delete_adapters()` and also there seems to be no clear path to completely unload a specific adapter. 

WDYT?

**Comment by BenjaminBossan** [comment]
> I wonder if it makes sense to add `adapter_names` to `unload_lora_weights()`. Users can always just use `delete_adapters()` and also there seems to be no clear path to completely unload a specific adapter.

So if a user passes adapter names to `unload_lora_weights()`, those adapters would be removed but the LoRA layer would stay? What if they pass all loaded adapters? In the end, do we not have the same open question there as with `delete_adapters`?

**Comment by sayakpaul** [comment]
> So if a user passes adapter names to unload_lora_weights(), those adapters would be removed but the LoRA layer would stay?

Yes, I guess we don’t have any other option here no?

> What if they pass all loaded adapters?

If they want to completely unload, then we already have `unload_lora_weights()`. 

> In the end, do we not have the same open question there as with delete_adapters?

Sorry, what was the open question?

**Comment by BenjaminBossan** [comment]
> Yes, I guess we don’t have any other option here no?

Yes, I don't think anything else would make sense in this situation.

> If they want to completely unload, then we already have `unload_lora_weights()`.

Okay, so the semantics would be: 

1. `unload_lora_weights(adapter_names)` unloads these specific adapters but keeps the LoRA layers, even if all adapters are removed.
2. `unload_lora_weights()` unloads all adapters and removes the LoRA layers.

I guess it's fine, if documented well.

> Sorry, what was the open question?

I meant what we just discussed, i.e. what happens if all adapters are passed to `adapter_names`.

**Comment by sayakpaul** [comment]
Thanks for the clarification!

> unload_lora_weights(adapter_names) unloads these specific adapters but keeps the LoRA layers, even if all adapters are removed.

I think this is already achievable with `delete_adapters()` and is sufficiently documented well:
https://huggingface.co/docs/diffusers/api/loaders/peft#diffusers.loaders.PeftAdapterMixin.delete_adapters

But we could document it in other guides too (cc: @stevhliu):

* https://huggingface.co/docs/diffusers/main/en/tutorials/using_peft_for_inference
